// cmnjs/date/format.js
// date - values that pass parse function or empty
// escape only used letters
// format('y-m-d h:i:s.u \\s\\upa \\d\\upa', new Date())

var parse = require('./parse');

function format(date, string) {

  date = parse(date) || new Date();

  if (!string) {
    string = 'y-m-d h:i:s.u';
  }

  return string.replace(/(\\?)(.)/g, function (useLess, escapeUsed, chr) {
    return (escapeUsed === '' && format.methods[chr])
      ? format.methods[chr](date) : chr;
  });

}

// export default methods, so they can be changed
format.methods = {

  y: function (d) {
    return d.getFullYear();
  },
  m: function (d) {
    return ('0' + (d.getMonth() + 1)).slice(-2);
  },
  d: function (d) {
    return ('0' + d.getDate()).slice(-2);
  },

  h: function (d) {
    return ('0' + d.getHours()).slice(-2);
  },
  i: function (d) {
    return ('0' + d.getMinutes()).slice(-2);
  },
  s: function (d) {
    return ('0' + d.getSeconds()).slice(-2);
  },
  u: function (d) {
    return ('00' + d.getMilliseconds()).slice(-3);
  }

};

module.exports = format;
