// @ali/alimusic-orbit-ejs-loader/index.js
var template    = require('lodash/template');
var minify      = require('html-minifier').minify;

var defaults = {
  engine: 'lodash',
  minify         : true,
  engineFull     : 'var _ = { escape: require(\'lodash.escape\') };',
  minifierOptions: {
    collapseInlineTagWhitespace: true,
    conservativeCollapse       : true,
    removeComments             : true,
    ignoreCustomFragments      : [/<%[\s\S]*?%>/]
  },
  templateOptions: {variable: 'data'}
};

module.exports = function (source) {

  this.cacheable && this.cacheable();

  var config = defaults;

  var templateLocal;

  if (config.engineFull === null) {
    config.engineFull = 'var _ = require(\'' + config.engine + '\');\n\n';
  }

  templateLocal = config.minify ? minify(source, config.minifierOptions) : source;

  // 删除由于是换行而多余的空格符号
  templateLocal = templateLocal.replace(/>\s+</g, '><');

  templateLocal = template(templateLocal, config.templateOptions);

  return config.engineFull + 'module.exports = ' + templateLocal + ';\n';

};
